core/pkg/eval/json_evaluator.go

```diff
@@ -243,54 +245,55 @@ func (je *JSONEvaluator) ResolveObjectValue(
 	value map[string]any,
 	variant string,
 	reason string,
+	metadata map[string]interface{},
 	err error,
 ) {
 	_, span := je.jsonEvalTracer.Start(ctx, "resolveObject")
 	defer span.End()
 
 	je.Logger.DebugWithID(reqID, fmt.Sprintf("evaluating object flag: %s", flagKey))
-	flag, _ := je.store.Get(flagKey)
-	return resolve[map[string]any](reqID, flagKey, context, je.evaluateVariant, flag.Variants)
+	return resolve[map[string]any](reqID, flagKey, context, je.evaluateVariant)
 }
 
-func resolve[T constraints](reqID string, key string, context map[string]any,
-	variantEval func(string, string, map[string]any) (string, string, error),
-	variants map[string]any) (
-	value T,
-	variant string,
-	reason string,
-	err error,
+func resolve[T constraints](reqID string, key string, context map[string]any, variantEval variantEvaluator) (
+	value T, variant string, reason string, metadata map[string]interface{}, err error,
 ) {
-	variant, reason, err = variantEval(reqID, key, context)
+	variant, variants, reason, metadata, err := variantEval(reqID, key, context)
 	if err != nil {
-		return value, variant, reason, err
+		return value, variant, reason, metadata, err
 	}
 
 	var ok bool
 	value, ok = variants[variant].(T)
 	if !ok {
-		return value, variant, model.ErrorReason, errors.New(model.TypeMismatchErrorCode)
+		return value, variant, model.ErrorReason, metadata, errors.New(model.TypeMismatchErrorCode)
 	}
 
-	return value, variant, reason, nil
+	return value, variant, reason, metadata, nil
 }
 
 // runs the rules (if defined) to determine the variant, otherwise falling through to the default
-func (je *JSONEvaluator) evaluateVariant(
-	reqID string,
-	flagKey string,
-	context map[string]any,
-) (variant string, reason string, err error) {
+func (je *JSONEvaluator) evaluateVariant(reqID string, flagKey string, context map[string]any) (
+	variant string, variants map[string]interface{}, reason string, metadata map[string]interface{}, err error,
+) {
+	metadata = map[string]interface{}{}
+
 	flag, ok := je.store.Get(flagKey)
 	if !ok {
 		// flag not found
 		je.Logger.DebugWithID(reqID, fmt.Sprintf("requested flag could not be found: %s", flagKey))
-		return "", model.ErrorReason, errors.New(model.FlagNotFoundErrorCode)
+		return "", flag.Variants, model.ErrorReason, metadata, errors.New(model.FlagNotFoundErrorCode)
```

Comment:
I think this won't work. since `ok` is `false` `flag` is empty

Reply:
Ya I think that's certainly true.